specification/dartLangSpec.tex

```diff
@@ -16231,238 +16306,445 @@ \subsection{Imports}
+to be used in more than one import,
+which means that the update of the binding of $p$
+does not interfere with other imports.%
 }
 
-\commentary{
-Note that no errors are raised if one hides or shows a name that is not in a namespace.
-}
-\rationale{
-This prevents situations where removing a name from a library would cause breakage of a client library.
+\LMHash{}%
+The effect of a repeated invocation of \code{$p$.loadLibrary()} is as follows:
+\begin{itemize}
+\item
+  If another invocation of \code{$p$.loadLibrary()} has already succeeded,
+  the repeated invocation also succeeds.
+  Otherwise,
+\item
+  If another invocation of \code{$p$.loadLibrary()} has failed:
+  \begin{itemize}
+  \item
+    If the failure is due to a compilation error,
+    the repeated invocation fails for the same reason.
+  \item
+    If the failure is due to other causes,
+    the repeated invocation behaves as if no previous call had been made.
+  \end{itemize}
+\end{itemize}
+
+\commentary{%
+In other words, a successful \code{loadLibrary()} guarantees that
+the import can be successfully accessed through the import prefix.
+If an invocation of \code{loadLibrary()} is initiated after
+another invocation has completed successfully,
+it is guaranteed to also complete successfully (success is idempotent).
+We do not specify which object the returned future resolves to.%
 }
+\EndCase
 
 \LMHash{}%
-The dart core library \code{dart:core} is implicitly imported into every dart library other than itself via an import clause of the form
+\Case{Semantics of immediate imports}
+When $I_i$ is an immediate import with a library URI, $u_i$,
+represented by the string $s_i$:
+Let $L_i$ be the library obtained from the source code denoted by $s_i$.
+We then say that the URI $u_i$ denotes the library $L_i$.
+All imports and exports of the same URI in a Dart program denotes
+the same library,
+and imports or exports of different URIs denote distinct libraries.
 
-\code{\IMPORT{} `dart:core';}
+\LMHash{}%
+Let \NamespaceName{\metavar{import}, i} be the namespace imported from $L_i$.
+The run-time namespace \NamespaceName{i} will then have the following bindings:
 
-unless the importing library explicitly imports \code{dart:core}.
+\begin{itemize}
+\item
+  For every top level function, getter, or setter $m$ named $n$ in
+  \NamespaceName{\metavar{import}, i},
+  a binding from $n$ to said function, getter, or setter.
+\item
+  For every name \id{} in \NamespaceName{\metavar{import}, i}
+  that is bound to a class, mixin, or type alias declaration
+  introducing a type $T$,
+  a binding from \id{} to the compiled representation of $T$.
+\end{itemize}
 
-\commentary{
-Any import of \code{dart:core}, even if restricted via \SHOW{}, \HIDE{} or \AS{}, preempts the automatic import.
-}
+\LMHash{}%
+If $I_i$ has prefix $p$,
+the run-time namespace of the current library
+maps $p$ to a non-deferred prefix run-time namespace \NamespaceName{p}
+containing the mappings of \NamespaceName{i}.
```

Comment:
Yes, NS_i is based on NS_import,i which is the namespace imported from L_i, and it has no conflicts (but it may have conflicted names among its bindings).